File: C++/BOJ/1012.cpp
#include<iostream>
#include<vector>

using namespace std;

void    dfs(vector<vector<int>> *table, int y, int x , int height, int width)
{
    if (y < 0 || y >= height || x < 0 || x >= width)
        return ;
    if ((*table)[y][x] == 0)
        return ;
    if ((*table)[y][x] == 1)
    {
        (*table)[y][x] = -1;
        dfs(table,y + 1, x ,height, width);
        dfs(table,y, x + 1 ,height, width);
        dfs(table,y - 1, x ,height, width);
        dfs(table,y, x - 1 ,height, width);
    }
}

int main()
{
    int t;
    cin >> t;
    for (int tc = 0; tc < t; tc++)
    {
        int m, n, cabbage_len;
        int x, y;
        int count = 0;
        cin >> m>> n>> cabbage_len;
        vector<vector<int>> table(n,vector<int>(m,0));

        for (int i = 0; i < cabbage_len; i++)
        {
            cin >> x>> y;
            table[y][x] = 1;
        }
        for (int i = 0; i < n; i++)
        {
            for (int j = 0; j < m; j++)
            {
                if (table[i][j] == 1)
                {
                    dfs(&table,i,j,n,m);
                    count++;
                }
            }
        }
        // for (int i = 0; i < table.size(); i++)
        // {
        //     for (int j = 0; j < table[i].size(); j++)
        //         cout <<table[i][j]<<" ";
        //     cout<<"\n";
        // }
        cout<<count<<"\n";
    }
}

File: Python/BOJ/1012.py
import sys
sys.setrecursionlimit(10**6)

def dfs(table, x, y, m, n):
    if x < 0 or y < 0 or x >= m or y >= n:
        return
    if table[y][x] == 0:
        return
    if table[y][x] == 1:
        table[y][x] = -1
        dfs(table, x, y + 1, m, n)
        dfs(table, x + 1, y, m, n)
        dfs(table, x, y - 1, m, n)
        dfs(table, x - 1, y, m, n)

def solution():
    t = int(sys.stdin.readline())
    for tc in range(t):
        count = 0
        m, n, cabbage_len = map(int, sys.stdin.readline().split())
        table = [[0 for _ in range(m)] for _ in range(n)]
        for cabbage in range(cabbage_len):
            x, y = map(int, sys.stdin.readline().split())
            table[y][x] = 1
        for i in range(n):
            for j in range(m):
                if table[i][j] == 1:
                    dfs(table,j,i,m,n)
                    count += 1
        print(count)

solution()
